central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,623 @@
+					RunState:                 storage.ReportStatus_WAITING,
+					ReportRequestType:        storage.ReportStatus_SCHEDULED,
+					ReportNotificationMethod: storage.ReportStatus_EMAIL,
+				},
+			},
+		}, false)
+		if err != nil {
+			log.Errorf("Error submitting scheduled report request for '%s': %s", reportConfig.GetName(), err)
+		}
+	}
+}
+
+func (s *scheduler) queuePendingReports(ctx context.Context) {
+	pendingReportsQuery := search.NewQueryBuilder().
+		AddExactMatches(search.ReportState, storage.ReportStatus_WAITING.String(), storage.ReportStatus_PREPARING.String()).
+		WithPagination(search.NewPagination().AddSortOption(search.NewSortOption(search.ReportQueuedTime))).
+		ProtoQuery()
+	pendingReports, err := s.reportMetadataStore.SearchReportMetadatas(ctx, pendingReportsQuery)
+	if err != nil {
+		log.Errorf("Error finding pending reports: %s", err)
+		return
+	}
+
+	for _, report := range pendingReports {
+		reportConfig, found, err := s.reportConfigDatastore.GetReportConfiguration(ctx, report.GetReportConfigId())
+		if err != nil {
+			log.Errorf("Error rescheduling pending report for report config ID '%s': %s", report.GetReportConfigId(), err)
+			continue
+		}
+		if !found {
+			log.Warnf("Report configuration with ID %s had pending reports but the configuration no longer exists",
+				report.GetReportConfigId())
+			continue
+		}
+		_, err = s.SubmitReport(&ReportRequest{
+			ReportConfig:   reportConfig,
+			ReportMetadata: report,
+		}, true)
+		if err != nil {
+			log.Errorf("Error rescheduling pending report for report config '%s': %s", report.GetReportConfigId(), err)
+		}
+	}
+}
+
+func (s *scheduler) queueScheduledReports(ctx context.Context) {
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportType, storage.ReportConfiguration_VULNERABILITY.String()).
+		ProtoQuery()
+	reportConfigs, err := s.reportConfigDatastore.GetReportConfigurations(ctx, query)
+	if err != nil {
+		log.Error("Error finding scheduled reports: %s", err)
+	}
+	for _, rc := range reportConfigs {
+		if rc.GetSchedule() != nil {
+			if err := s.UpsertReportSchedule(rc); err != nil {
+				log.Errorf("Error queuing scheduled report for report configuration with ID %s: %v", rc.GetId(), err)
+			}
+		}
+	}
+}
+
+/* Utility Functions */
+
+func (s *scheduler) doesUserHavePendingReport(ctx context.Context, configID string, userID string) (bool, error) {
+	s.dbLock.Lock()
+	defer s.dbLock.Unlock()
+
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportConfigID, configID).
+		AddExactMatches(search.ReportState, storage.ReportStatus_WAITING.String(), storage.ReportStatus_PREPARING.String()).
+		AddExactMatches(search.ReportRequestType, storage.ReportStatus_ON_DEMAND.String()).
+		ProtoQuery()
+	runningReports, err := s.reportMetadataStore.SearchReportMetadatas(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	for _, rep := range runningReports {
+		if rep.GetRequester().GetId() == userID {
+			return true, nil
+		}
+	}
```

Comment:
You should get this from the the in-mem running jobs

Reply:
I will address this one in a followup PR.